src/appleseed/foundation/platform/system.cpp

```diff
@@ -421,6 +429,388 @@ uint64 System::get_process_virtual_memory_size()
     return static_cast<uint64>(rss) * sysconf(_SC_PAGESIZE);
 }
 
+// ------------------------------------------------------------------------------------------------
+// FreeBSD.
+// ------------------------------------------------------------------------------------------------
+
+#elif defined __FreeBSD__ && (__x86_64__ || __i386__)
+
+// FreeBSD does not provide an API to query CPU cache information, so we'd
+// have to ask for it by ourselves...
+
+// Since we only want to know cache size and line size, structure below is
+// very simplistic.  Initially, all caches have zero size (i.e. absent).
+static struct trivial_x86_cache
+{
+    size_t size;
+    size_t linesize;
+} mycaches[3];
+
+enum { L1D, L2, L3 };
+enum { eax, ebx, ecx, edx };
+
+// %ebx may be used to point to GOT (Global Offset Table) for PIC (Position
+// Independent Code) on 32-bit x86, so it must be preserved.  Normally
+// compilers handle this implicitly because %ebx is also callee saved, but
+// GCC before 5.0 forbids any use of %ebx with PIC, so it must be performed
+// explicitly.  Unfortunately, we need a separate implementation for x86-64
+// to preserve %rbx because 32-bit operations would set the upper 32 bits
+// to zero.
+static inline void
+cpuid(uint32_t* regs)
+{
+    asm(
+#if __x86_64__
+        "movq %%rbx, %q1\n\t"
+        "cpuid\n\t"
+        "xchgq %%rbx, %q1"
+#else
+        "movl %%ebx, %1\n\t"
+        "cpuid\n\t"
+        "xchgl %%ebx, %1"
+#endif
+      : "+a" (regs[eax]),
+        "=r" (regs[ebx]),
+        "+c" (regs[ecx]),
+        "=d" (regs[edx]));
+}
+
+#define bit(n)          (1ul << (n))
+#define bitmask(h,l)    ((bit(h) | (bit(h) - 1)) & ~(bit(l) - 1))
+#define bitfield(x,h,l) (((x) & bitmask(h, l)) >> l)
+
+// For modern CPUs, we use Deterministic Cache Parameters (Function 04h) to
+// obtain cache information.
+static void
+get_cache_info_deterministic(struct trivial_x86_cache *caches)
+{
+    uint32_t regs[4];
+
+    // Cycle up to ten possible caches to be extra sure.
+    for (size_t i = 0; i < 10; i++)
+    {
+        regs[eax] = 4;
+        regs[ecx] = i;
+        cpuid(regs);
+
+        unsigned type = bitfield(regs[eax], 4, 0);
```

Comment:
We usually tag with `const` all variables whose value don't change after initialization.